### ISSUE TITLE
Summary parsing

### ISSUE THREAD

**Issue by KurtzPT** [ZeroQI/Hama.bundle#285]
For the anime Air Gear, Hama is getting the following summary from AniDB:
![image](https://user-images.githubusercontent.com/14930057/53690687-429caa00-3d67-11e9-875d-e993548b27d7.png)

It does not remove the line that starts with *, nor does it remove the last line ("Summary:")

**Comment by ZeroQI** [comment]
Hama is not editing summaries currently.
it can be done at source level, or at processing level but moifying at Source (AniDB) is cleaner as i like the engine simpler. adding to to-do list

**Comment by KurtzPT** [comment]
There's a function on AniDB.py that strips some stuff:

> def summary_sanitizer(summary):
  summary = summary.replace("`", "'")                                                                # Replace backquote with single quote
  summary = re.sub(r'https?://anidb\.net/[a-z]{1,2}[0-9]+ \[(?P<text>.+?)\]', r'\g<text>', summary)  # Replace links
  summary = re.sub(r'^\* B.*\n+|\nSource:\w*[\w\W]*|\nNote:\w*[\w\W]*()',     "",          summary)  # Remove Source and M=Notes
  return summary

**Comment by ZeroQI** [comment]
Excellent. Need to add to that replacement of
- "* Adapted..."
- "Summary: ..."

```
def summary_sanitizer(summary):
  summary = summary.replace("`", "'") # Replace backquote with single quote
  summary = re.sub(r'https?://anidb.net/[a-z]{1,2}[0-9]+ [(?P.+?)]', r'\g', summary) # Replace links
  summary = re.sub(r'^* \B\n',                                       "",    summary) #* Adapted ...
  summary = re.sub(r'\nSource: \w[\w\W]',                            "",    summary) #Source:
  summary = re.sub(r'\nNote: \w[\w\W]*',                             "",    summary) #Note: 
  summary = re.sub(r'\nSummary: \w[\w\W]*',                          "",    summary) #Summary:
  return summary
```

**Comment by EndOfLine369** [comment]
Added regex enhancement to existing open pull request. We were also not cleaning the episode summaries so we are now doing that.

**Comment by EndOfLine369** [comment]
@KurtzPT, all good? you mentioned you updated post merge in other ticket so you should be able to test.

**Comment by KurtzPT** [comment]
@EndOfLine369 almost good. The example I gave for the specials mapping (https://anidb.net/perl-bin/animedb.pl?show=anime&aid=5629) it removes the line that starts with * but it leaves (I think) new line character:

![image](https://user-images.githubusercontent.com/14930057/53874626-b4405680-3ffa-11e9-8c9d-634618531692.png)

Editing manually the summary to remove the new line:

![image](https://user-images.githubusercontent.com/14930057/53874678-ce7a3480-3ffa-11e9-9869-3486479b27ef.png)

**Comment by sven-7** [comment]
If cleaning episode descriptions from AniDB, many of them have the last line as "Source: Crunchyroll" or "Source: amazon".  I imagine there are some others like Funimation, Hulu, Netflix, etc. 

CR:  http://anidb.net/perl-bin/animedb.pl?show=ep&eid=199578
Amazon:  http://anidb.net/perl-bin/animedb.pl?show=ep&eid=190903

**Comment by EndOfLine369** [comment]
@KurtzPT fixed

@sven-7 what is the question your asking?
All text after "(Source|Note|Summary):" is removed.

**Comment by sven-7** [comment]
@EndOfLine369 - okay, I missed that post. My bad.

**Comment by EndOfLine369** [comment]
nw @sven-7, made it a vague as I could so we don't constantly have to update for different wordings.
```python
  summary = re.sub(r'^\* .*',                     "",      summary, flags=re.MULTILINE)  # Remove the line if it starts with '* '
  summary = re.sub(r'\n(Source|Note|Summary):.*', "",      summary, flags=re.DOTALL)     # Remove all lines after this is seen
```

**Comment by sven-7** [comment]
Makes sense to me - that should keep it super easy. 

Here are some other ones I've seen have [multiple](http://anidb.net/perl-bin/animedb.pl?show=anime&aid=4576) "notes" in the descriptions. I think you'll get 99.9% of them, but I did see one that had the below in doing some random searches on AniDB:

- [— written by Singan](http://anidb.net/perl-bin/animedb.pl?show=anime&aid=10369)
- [~ translated by](http://anidb.net/perl-bin/animedb.pl?show=anime&aid=7971)

**Comment by purposelycryptic** [comment]
Waaaay back when I had actual free time and wasn't sick as a did with Pneumonia, I added this functionality to my fork (along with pulling AniDB Average Ratings vs Weighted Ratings), intending to eventually see if I could merge them when I got a bit further. 

Then all the stuff with getting specials properly recognized/organized got way more complicated than it seemed at first, and @EndOfLine369 started contributing more code than my brain could parse (that is a very good thing, just overwhelming), Hama got large chunks rewritten... and then I shattered my middle finger in my dominant hand, and disappeared from the web for several months while getting surgery, then healing up. Now I've been such in bed with pneumonia for over a month... Not the best 6 months ever.

ANYWAY, just to the point, the code is still in my fork, should be really easy to find given the small handful of commits. It's been thoroughly tested on my full anime library (1,693 series and growing), and all the "*based on..., "*Summary written by...", "*received really screening at... on...", etc lines are gone from the series summaries, and the "*source: Crunchyroll", etc lines from the episode summaries. It actually took me quite a while to get the RegEx just right, because RegEx is a pain at times. The functions should actually bee commented and named (something like summary cleaner, episode summary cleaner).

SO, and my apologies for the rambling (still have a fever, kind of dizzy), in case you haven't gotten it working 100% of the time already and I just missed it, I just thought I'd mention it, in case you want to use it. I fidlled with some of the code a little that might not have made a commit, but it shouldn't have involved that in any way.

I'm actually still using a version of Hama from back when we were trying to get the specials numbering to line up with AniDB, with the changes from my fork integrated, because something clicked in the code, and it has just been working flawlessly ever since, and I can't think of any way for it to work any more perfectly.

Again, sorry for the rambling, between the meds and the sickness, my head is in pretty sorry shape. Just happened to notice the GitHub status update email, and thought it might be worth mentioning.

On another not, thanks for all your hard work, EOL, ZeroQI, and I'm really sorry for disappearing back then - I had to have 3 separate surgeries on the joint in my finger, pulling out bone-splinters, reattaching ligaments, etc, and was on a lot of pain meds for quite a while, essentially in zombie-mode for ages.

**Comment by EndOfLine369** [comment]
@purposelycryptic, don't let my number of commits put you off from reading/understanding what is changed 😃. I favor a commit per change instead of one commit containing 50 changes. makes it more readable what exactly what was done to do the change and also if it did break something it is easier to undo. I recommend the GitHub Desktop view of the repo makes it easier to review the changes over having to click through the website for each commit. You also have no fork of this repo [purposelycryptic?tab=repositories](https://github.com/purposelycryptic?tab=repositories)



@sven-7 / @ZeroQI, I could change it to:
```
summary = re.sub(r'^(\*|--|~) .*',                     "",      summary, flags=re.MULTILINE)  # Remove the line if it starts with ('* ' / '-- ' / '~ ')
```
should I? Are those two additions `( '^-- .*' & '^~ .*' )` be unique enough to always be not part of the actual summary?

**Comment by ZeroQI** [comment]
@EndOfLine369 I believe so.
@purposelycryptic take care of yourself. Will try to take back a more active role

**Comment by purposelycryptic** [comment]
Thanks, @ZeroQI  :-)

@EndOfLine369 :
my memories must be playing tricks on me in my old age (and lying in bed feverish with Pneumonia definitely doesn't help). I was also working on a ShokoServer fork in parallel at the same time, similar issues, and it looks like I was stupid enough to never actually make a fork of Hama on GitHub, just locally editing it. It looks like I didn't push some of my later updates to ShokoServer either... Really sloppy :-/

Your nicely documented per-change commits are what what we all should aspire to (Especially me ;-) )

I snuck out of bed (well, more slowly dragged myself, technically, considering I live alone), and looked through my files, finally just checking the ones currently in use by Plex, but I can't help but feel like I'm missing something.

My summary_sanitizer function definitely looks a good bit different, though (more messy, too), though, since my version is based on commit [fb20b12b](https://github.com/ZeroQI/Hama.bundle/commit/fb20b12bbe51178395d560c913575984349361b4) from last August, it would be stranger not to see everything look different.

Here is what my summary_sanitizer looks like:
```python
def summary_sanitizer(summary):
	summary = summary.replace("`", "'")                                                      # Replace backquote with single quote
	summary = re.sub(r'http://anidb\.net/[a-z]{1,2}[0-9]+ \[(.+?)\]',       r'\1', summary)  # Replace links
	summary = re.sub(r'^\* B.*\n+|\nSource:\w*[\w\W]*|\nNote:\w*[\w\W]*()', "",    summary)  # Remove Source and M=Notes
	 return summary
```

If it wasn't identical to the one in my ShokoServer fork (also in need of a commit on GH to catch up), except for that being in C#, I would swear there was more, but maybe my melted brain just isn't up to the task.

I did just finally fork HAMA, and added my changes, namely the alternate summarySanitizer, and pulling the average vs weighted AniDB ratings. [Take a look](https://github.com/purposelycryptic/Hama.bundle), if you're interested (or bored). Hopefully I really didn't miss anything (it's just _so little_ code!), but if I did, I blame being sick as a dog, and will hopefully add what is missing when I'm back on my feet. Maybe I was thinking of ASS?

As always, thanks for keeping up the good work - if that sounds like a generic platitude, it isn't. HAMA is the reason I switched from Emby after almost a decade of use, and the best thing about Plex :-D

Now, back to bed before I die...

**Comment by purposelycryptic** [comment]
> @purposelycryptic, don't let my number of commits put you off from reading/understanding what is changed . I favor a commit per change instead of one commit containing 50 changes. makes it more readable what exactly what was done to do the change and also if it did break something it is easier to undo. I recommend the GitHub Desktop view of the repo makes it easier to review the changes over having to click through the website for each commit. You also have no fork of this repo [purposelycryptic?tab=repositories](https://github.com/purposelycryptic?tab=repositories)
> 
> @sven-7 / @ZeroQI, I could change it to:
> 
> ```
> summary = re.sub(r'^(\*|--|~) .*',                     "",      summary, flags=re.MULTILINE)  # Remove the line if it starts with ('* ' / '-- ' / '~ ')
> ```
> should I? Are those two additions `( '^-- .*' & '^~ .*' )` be unique enough to always be not part of the actual summary?

I can't remember what they were, but while testing I remember finding a few series where the actual summary started with a '*', but none with a '*' followed by a space before text (or something equally weird). 

Sorry if that isn't much help, too much stuff happened between then and now, and my memory isn't what it used to be.

**Comment by EndOfLine369** [comment]
@purposelycryptic, that was the logic we already had that was not good enough https://github.com/ZeroQI/Hama.bundle/commit/ae3a8b4efe41243f219772525aaacd01ee67ebad .

@ZeroQI, updated https://github.com/ZeroQI/Hama.bundle/commit/903adbdbda350462a078fd6f67a3ff984c43cb27 .

**Comment by purposelycryptic** [comment]
@EndOfLine369: I figured. There was something else, that RegEx alone couldn't account for all the surplus lines removed, no matter how many times I looked at it. The problem is what, and where.

*sigh* maybe it's just been too long, and I should just count my blessings that it works as well as it does. Hunting through old code with a high fever was probably never the best idea.

Sorry for ending up wasting everyone's time.

**Comment by ZeroQI** [comment]
@purposelycryptic Only the ones that dare do something can get it wrong, no need to  be sorry
I am closing this, but feel free to comment if i was wrong in any way and it wasn't resolved.

**Comment by purposelycryptic** [comment]
@ZeroQI: Thanks for the encouragement:-)
If everything is working now, that's the best outcome I could hope for - as my Plex Server is finally working almost exactly the way I want it, I'm afraid of touching anything for fear of upsetting the delicate balance, so I can't really try it out, since I REALLY shouldn't leave bed, bad Pneumonia and all; my test boxes are at the other end of the house (as is my server, but I can get on it over Steam Link if it doesn't randomly stop). Mostly been living off of my little Android TV box.

So, this may or may not going to be of any use, but I noticed that every version of a Summary Cleaner I have (In HAMA, In Shoko, and in ShokoMetadata - all forks) each uses slightly different RegEx for some reason. So, just in case any of it proves worthwhile someday, I figured I'd at least leave the code snippets here for posterity. Btw - I finally thought of why the HAMA function I posted above is identical to what you had - I had completely forgotten about [this](https://github.com/ZeroQI/Hama.bundle/issues/232) lol

From [ShokoMetadata.bundle](https://github.com/purposelycryptic/ShokoMetadata.bundle)
The Commit was titled "**_Updated updated code to actually work_**" It was posted after you had already committed the cleaned up version of the barely readable one-liner I posted, so **>shrug<**. Not sure if the minor changes were actually of benefit or not.
```python
#Strip internal AniDB links that Plex can't use, as well as "Based on:" and Initial airing notes and Summary Source citations from summary.
metadata.summary = re.sub(r'http://anidb\.net/[a-z]{1,2}[0-9]+ \[(.+?)\]', r'\1', re.sub(r'\* B.*\n+|\nSource:[\w\W]*|\nNote:[\w\W]*', "", try_get(series, 'summary'))).replace("`", "'")
```
It also has a separate episode summary cleaner function:
```python
#Remove annoying "Episode Overview not Available" from episodes with no summary
if (ep['summary'] != "Episode Overview not Available"): 
episodeObj.summary = ep['summary']
#Remove "Source:..." line frequently found in AniDB Episode Summaries
episodeObj.summary = re.sub(r'\nSource:[\w\W]*', "", ep['summary']) 
```

From [ShokoServer](https://github.com/purposelycryptic/ShokoServer)
Shoko is written in C#, hence the different code. This commit was about a month after the one above, and again, a small modification... Pretty sure the additional "A" line killer was meant for summaries that either started with "Adaptation of ...", or for those that ended with something like "Adapted from MAL by..."
```cs
//anime.Description = TryGetProperty(docAnime, "anime", "description")?.Replace('`', '\'');
//Replaced above with cleaned description
string desc = TryGetProperty(docAnime, "anime", "description")?.Replace('`', '\'');
string linkRemoveRE = @"http://anidb\.net/[a-z]{1,2}[0-9]+ \[(.+?)\]";
string noteRemoveRE = @"\* B.*\n+|\* A.*\n+|\nSource:[\w\W]*|\nNote:[\w\W]*";
string replaceRE = "$1";
desc = Regex.Replace(desc, linkRemoveRE, replaceRE);
desc = Regex.Replace(desc, noteRemoveRE, "");
anime.Description = desc;
```

So, as I said, this may or may not be of any use, but at least it's here if it is. Now, back to bed before I die ;-)
